### PR TITLE
Ftr: add protobuf support

### PR DESCRIPTION
**Pull request by skyitachi**
**What this PR does**:
- support protobuf
- some code refactoring, add `Serializer` interface and serialization layer

**What this PR fix**:
- registry url should not be encoded twice in `zookeeper/registry.go`

**Special notes for your reviewer**: file changes maybe too much, please be patient

**Does this PR introduce a user-facing change?**: No, all things are compatible.

ps: protoc-gen-dubbogo plugin are in the [repo](https://github.com/skyitachi/protoc-gen-dubbogo)

### CODE REVIEW

**Comment by AlexStocks** [comment]
@skyitachi pls check travis result.

File: protocol/dubbo/const.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
+package dubbo
+
+import (
+	"github.com/pkg/errors"
```
Comment (AlexStocks) [inline protocol/dubbo/const.go]: change it to another import block.

File: protocol/dubbo/const.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+	STRING_SHORT_MAX  = 0x3ff
+
+	BC_TRUE = byte('T')
+
+	P_PACKET_CHUNK = byte(0x4f)
+	P_PACKET       = byte('P')
+
+	P_PACKET_DIRECT   = byte(0x80)
+	PACKET_DIRECT_MAX = byte(0x7f)
+
```
Comment (AlexStocks) [inline protocol/dubbo/const.go]: hey guy, the above constants have been defined in dubbo-go-hessian2.

File: protocol/dubbo/const.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	// magic header
+	MAGIC      = uint16(0xdabb)
+	MAGIC_HIGH = byte(0xda)
+	MAGIC_LOW  = byte(0xbb)
+
+	// message flag.
+	FLAG_REQUEST = byte(0x80)
+	FLAG_TWOWAY  = byte(0x40)
+	FLAG_EVENT   = byte(0x20) // for heartbeat
+	SERIAL_MASK  = 0x1f
```
Comment (AlexStocks) [inline protocol/dubbo/const.go]: the same as above, u should use dubbo-go-hessian2  constants instead.

File: protocol/dubbo/proto.go [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+		} else {
+			responseType = RESPONSE_WITH_EXCEPTION
+		}
+	}
+	// write response type
+	if err := writeByte(buf, responseType); err != nil {
+		return nil, err
+	}
+	if response.Exception != nil {
+		// 简单处理一下异常
```
Comment (AlexStocks) [inline protocol/dubbo/proto.go]: translate it to english.

File: protocol/dubbo/proto.go [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+func readObject(reader io.Reader, value proto.Message) error {
+	_, err := pbutil.ReadDelimited(reader, value)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// 对应java protobuf serialize 实现
```
Comment (AlexStocks) [inline protocol/dubbo/proto.go]: translate it to english.

File: protocol/dubbo/proto.go [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+	}
+	*value = i32v.Value
+	return nil
+}
+
+//
+func readDelimitedLength(reader io.Reader, length *int) error {
+	var headerBuf [binary.MaxVarintLen32]byte
+	var bytesRead, varIntBytes int
+	var messageLength uint64
```
Comment (AlexStocks) [inline protocol/dubbo/proto.go]: in dubbo-go, we always put some blocks in a var () block.

**Comment by AlexStocks** [comment]
@hxmhlt @fangyincheng pls check this pr. the author has changed so many dubbo package files.

File: protocol/dubbo/client.go [hunk 7]
```diff
@@ -240,6 +224,37 @@ func (c *Client) call(ct CallType, request *Request, response *Response, callbac
...
+	svc.Path = strings.TrimPrefix(request.svcUrl.Path, "/")
+	svc.Interface = request.svcUrl.GetParam(constant.INTERFACE_KEY, "")
+	svc.Version = request.svcUrl.GetParam(constant.VERSION_KEY, "")
+	svc.Group = request.svcUrl.GetParam(constant.GROUP_KEY, "")
+	svc.Method = request.method
+	svc.Timeout = c.opts.RequestTimeout
+	p := NewClientRequestPackage(header, svc)
+
+	serialization := request.svcUrl.GetParam("serialization", c.conf.Serialization)
+	if serialization == constant.HESSIAN2_SERIALIZATION {
```
Comment (hxmhlt) [inline protocol/dubbo/client.go]: request.svcUrl.GetParam("serialization", c.conf.Serialization)
=>
request.svcUrl.GetParam(SERIALIZATION_KEY, c.conf.Serialization)

File: protocol/dubbo/proto.go [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+ */
+
+package dubbo
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"github.com/apache/dubbo-go/common/constant"
+	"io"
```
Comment (hxmhlt) [inline protocol/dubbo/proto.go]: import fmt

Comment (hxmhlt) [inline protocol/dubbo/proto.go]: import fmt

File: common/extension/serialization.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+		constant.S_Proto:    constant.PROTOBUF_SERIALIZATION,
+	}
+}
+
+func SetSerializer(name string, serializer interface{}) {
+	serializers[name] = serializer
+}
+
+func GetSerializer(name string) interface{} {
+	return serializers[name]
```
Comment (hxmhlt) [inline common/extension/serialization.go]: Why not return Serializer type other than interface{}?

File: protocol/dubbo/const.go (hunk 2 above)
Comment (skyitachi) [reply protocol/dubbo/const.go]: I know，but dubbo-go-hessian2 do a lot more work other than serialize include dubbo codec，so I use dubbo codec directly and go-hessian2 as  pure serializer，just as what does in dubbo java，keep go-hessian2’s serialize functions

File: protocol/dubbo/const.go (hunk 3 above)
Comment (skyitachi) [reply protocol/dubbo/const.go]: same as above

File: common/extension/serialization.go (hunk 9 above)
Comment (xujianhai666) [reply common/extension/serialization.go]: I found this method is extra, maybe we should delete ?

File: common/extension/serialization.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+package extension
+
+import (
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/pkg/errors"
+)
+
+var (
+	serializers = make(map[string]interface{})
```
Comment (xujianhai666) [inline common/extension/serialization.go]: maybe we should abstract a interface

File: protocol/dubbo/codec.go [hunk 11]
```diff
@@ -46,64 +61,254 @@ const (
...
+		}
+	case PackageRequest_TwoWay:
+		bs = append(bs, hessian.DubboRequestHeaderBytesTwoWay[:]...)
+	}
+	bs[2] |= header.SerialID & hessian.SERIAL_MASK
+	binary.BigEndian.PutUint64(bs[4:], uint64(header.ID))
+	return bs
 }
 
-func (p *DubboPackage) Marshal() (*bytes.Buffer, error) {
```
Comment (xujianhai666) [inline protocol/dubbo/codec.go]: why not use Marshal method name

File: protocol/dubbo/codec.go [hunk 12]
```diff
@@ -46,64 +61,254 @@ const (
...
+			return errors.WithStack(err)
+		}
+		rsp, ok := p.Body.(*ResponsePayload)
+		if !ok {
+			return errors.Errorf("java exception:%s", exception.(string))
+		}
+		rsp.Exception = errors.Errorf("java exception:%s", exception.(string))
+		return nil
+	} else if p.IsHeartBeat() {
+		// heartbeat
```
Comment (xujianhai666) [inline protocol/dubbo/codec.go]: no  chinese comment here, pls remove

File: protocol/dubbo/codec.go [hunk 13]
```diff
@@ -46,64 +61,254 @@ const (
...
+	// request id
+	binary.BigEndian.PutUint64(byteArray[4:], uint64(header.ID))
+
+	// body
+	body, err := serializer.Marshal(p)
+	if err != nil {
+		return nil, err
+	}
+
+	//byteArray = encNull(byteArray) // if not, "java client" will throw exception  "unexpected end of file"
```
Comment (xujianhai666) [inline protocol/dubbo/codec.go]: is that extra?

File: protocol/dubbo/codec_test.go [hunk 14]
```diff
@@ -57,18 +65,133 @@ func TestDubboPackage_MarshalAndUnmarshal(t *testing.T) {
...
+
+	pkgres = NewDubboPackage(data)
+	pkgres.SetSerializer(ProtoSerializer{})
+	err = pkgres.Unmarshal()
+	assert.NoError(t, err)
+	body, ok := pkgres.Body.(map[string]interface{})
+	assert.Equal(t, ok, true)
+	req, ok := body["args"].([]interface{})
+	assert.Equal(t, ok, true)
+	// protobuf的rpc参数只能是一个
```
Comment (xujianhai666) [inline protocol/dubbo/codec_test.go]: chinese comment， u know

File: protocol/dubbo/const.go (hunk 3 above)
Comment (xujianhai666) [reply protocol/dubbo/const.go]: I thought we should discuss with @AlexStocks, I found hessian codec impl below.

**Comment by xujianhai666** [comment]
@skyitachi I thought we should discuss with @AlexStocks . If we should move hessian from hessian repo to here, we should seperate those two work.

File: protocol/dubbo/codec.go (hunk 11 above)
Comment (skyitachi) [reply protocol/dubbo/codec.go]: Marshal Method in the `package.go` file

File: common/extension/serialization.go (hunk 10 above)
Comment (skyitachi) [reply common/extension/serialization.go]: I have written a `Serializer` interface in `dubbo/serialize.go`,  use `interface{}` here becase of extension need to import something that in dubbo directory (which may cause import cycles), I use extension in `dubbo`, so it's just work around

**Comment by AlexStocks** [comment]
> @skyitachi I thought we should discuss with @AlexStocks . If we should move hessian from hessian repo to here, we should seperate those two work.

u should have a discuss with @hxmhlt

**Comment by AlexStocks** [comment]
@xujianhai666 any improvement progress recently?

**Comment by hxmhlt** [comment]
@skyitachi Pls contact me as soon as possible.

**Comment by hxmhlt** [comment]
I suggest that creating a base protocol like 'BaseTCPProtocol' which  inherited by dubboProtocol & protobufProtocol is more better .

**Comment by skyitachi** [comment]
> @skyitachi Pls contact me as soon as possible.

I think should make the pr separately, first is decouple dubbo codec and hessian serializer

**Comment by skyitachi** [comment]
I think should make a new pr :)

**Pull request closed without merging**